clients/src/main/java/org/apache/kafka/common/errors/FatalExitError.java

```diff
@@ -0,0 +1,77 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more contributor license agreements. See the NOTICE
+ * file distributed with this work for additional information regarding copyright ownership. The ASF licenses this file
+ * to You under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the
+ * License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package org.apache.kafka.common.errors;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The exception that indicates the need to exit the process. This exception is expected to be
+ * caught at the highest level of the thread so that no shared lock would be held by the thread
+ * when it calls {@link System#exit(int)}
+ */
+public class FatalExitError extends Error {
+
+    private static final Logger log = LoggerFactory.getLogger(FatalExitError.class);
+
+    private final static long serialVersionUID = 1L;
+
+    protected int exitStatus;
+
+    public FatalExitError(int exitStatus, String message, Throwable cause) {
+        super(message, cause);
+        this.exitStatus = exitStatus;
+    }
+
+    public FatalExitError(int exitStatus, String message) {
+        super(message);
+        this.exitStatus = exitStatus;
+    }
+
+    public FatalExitError(int exitStatus, Throwable cause) {
+        super(cause);
+        this.exitStatus = exitStatus;
+    }
+
+    public FatalExitError(int exitStatus) {
+        super();
+        this.exitStatus = exitStatus;
+    }
+
+    public void shutdownSystem() {
```

Comment:
Thanks for the review. Yeah I also was thinking that it also could be moved to Utils. I don't have a strong position of where it should it be put.  Can you clarify a bit more about moving `shutdownSystem` to `kafka.Kafka.main()`? Are you suggesting to have the method `shutdownSystem` inside the method `kafka.Kafka.main()`? 

That is true. Specially that current code base is a bit liberal in catching Throwable and to have the new way of exiting work we need to either i) remove all those catch throwable clauses (which could also change the way the code works), ii) exclude FatalExitError from those clauses to make sure that it is not caught. That latter seems more safe. An alternative is to gradually migrate from the current model to the proposed Error-based model and to limit the current patch to only the one case in `ReplicaFetcherThread`.

I am still not sure what logic you are suggesting to be moved to the `kafka.Kafka.main()` function. If threads do not catch an Error (which they should not) the thread would not continue running and thus exits. I am not sure how moving some logic to the main function could help that.